### PR TITLE
Implement Model Checkpointing

### PR DESCRIPTION
**Pull request by ben-heil**
Main Ideas:
1. Implement logic to store trained models and use them later
2. Add the ability to select which disease should be used to train the model

Changes:
- Creates `test_model_saving.ipynb` to ensure that everything is working as expected
- Adds `test_generate_encoding` to make sure the one-hot encoding function in `util.py` works
- Have the training functions in `classifier.py` check whether the dataset corresponding to each study actually has any data in it. If we're training on tb and the study only has sepsis samples, we don't need a dataset object for it.
- Add model saving logic to training functions in `classifier.py`
- Modify `dataset.py` to have datasets detect when they don't have any useful data 
- Create `generate_encoding` function in utils to make one-hot encodings for lists of classes
- Create `keep_samples_with_labels` function to act as the inverse of the `remove_samples_with_labels` function in `util.py`

### CODE REVIEW

File: analyses/test_model_saving.ipynb [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    "logger = logging.getLogger(__name__)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Ensure the models train deterministically\n",
```
Comment (ajlee21) [inline analyses/test_model_saving.ipynb:62]: Did you re-run your training to see if you did in fact get the similar model (model weights) each time?  I remember that I had trouble trying to train my VAE multiple times because the weights learned were slightly different since there is some randomness to the learning process that can't be helped

File: analyses/test_model_saving.ipynb [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "irm_results = classifier.train_with_irm(net, map_file, train_dirs, tune_dirs, gene_file, num_epochs, \n",
```
Comment (ajlee21) [inline analyses/test_model_saving.ipynb:144]: What is IRM?

File: analyses/test_model_saving.ipynb [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+   "source": [
+    "print('Trained network tune accuracy: {}'.format(tune_acc))\n",
+    "print('Trained network tune loss: {}'.format(avg_loss))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Test untrained network"
```
Comment (ajlee21) [inline analyses/test_model_saving.ipynb:232]: What are the default weights for this untrained model?

Comment (ajlee21) [reply analyses/test_model_saving.ipynb:232]: What is the purpose of doing this?

File: test/test_utils.py [hunk 4]
```diff
@@ -41,6 +43,27 @@ def test_remove_samples_with_label(df, sample_to_label, label_to_remove, correct
...
+class3_encoding = np.zeros((3, 3))
+class3_encoding[2, 2] = 1
+
+
+@pytest.mark.parametrize('classes,correct_encodings',
+                         [(['class1', 'class2', 'class3'],
+                           [class1_encoding, class2_encoding, class3_encoding]),
+                          (['class1', 'class2'], [0, 1]),
+                          ])
+def test_generate_encoding(classes, correct_encodings):
```
Comment (ajlee21) [inline test/test_utils.py:60]: Maybe add a comment for what you are testing here.

File: whistl/classifier.py [hunk 5]
```diff
@@ -30,7 +30,8 @@ def compute_irm_penalty(loss, dummy_w):
 
 
 def train_with_irm(classifier, map_file, train_dirs, tune_dirs, gene_file,
-                   num_epochs, loss_scaling_factor, label_to_encoding, device, logger=None):
+                   num_epochs, loss_scaling_factor, label_to_encoding, device, logger=None,
+                   save_file=None, burn_in_epochs=100):
     '''Train the provided classifier using invariant risk minimization
     IRM looks for features in the data that are invariant between different environments, as
```
Comment (ajlee21) [inline whistl/classifier.py:36]: Just out of curiosity, what do you mean by invariant here?

**Review by ajlee21 — APPROVED** [review]
Very nice PR.  

Nicely commented and organized.

File: analyses/test_model_saving.ipynb (hunk 1 above)
Comment (ben-heil) [reply analyses/test_model_saving.ipynb:62]: I have previously, and I got the same loss if I remember correctly. The trick is to set the backend to deterministic mode.

File: analyses/test_model_saving.ipynb (hunk 2 above)
Comment (ben-heil) [reply analyses/test_model_saving.ipynb:144]: IRM is invariant risk minimization. It's a method of regularization that penalizes models that don't perform well across different environments. In this case, the environments are the different experiments.

File: analyses/test_model_saving.ipynb (hunk 3 above)
Comment (ben-heil) [reply analyses/test_model_saving.ipynb:232]: The weights are initialized with the default PyTorch scheme, which I think is uniform with a range proportional to the square root of the number of weights in the layer.

I ran the untrained model as a negative control. If an untrained model had good performance at classifying the test data for some reason, it would mean that the loading the trained model wasn't necessarily successful.

File: test/test_utils.py (hunk 4 above)
Comment (ben-heil) [reply test/test_utils.py:60]: Sounds good

File: whistl/classifier.py (hunk 5 above)
Comment (ben-heil) [reply whistl/classifier.py:36]: I may have already explained this [earlier](https://github.com/greenelab/whistl/pull/13#discussion_r350779490), but they are invariant in the sense that the model doesn't have more or less error in different environments. For more information on the details of how that is accomplished, see https://arxiv.org/abs/1907.02893
